sdk/nodejs/x/automation/cmd.ts

```diff
@@ -0,0 +1,83 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as childProcess from "child_process";
+
+// TODO the error design here needs some work.
+// need to investigate the state of the art in structured promise error handling in nodejs libs
```

Comment:
The typical way of doing this in JS/TS is to throw an `Error` (i.e. `reject(new Error(...))`).

If we want to include more information along with the error, we can extend `Error`, something like:

```typescript
export class CommandError extends Error {
    constructor(message: string, readonly result: CommandResult) {
        super(message)
        this.name = "CommandError"
    }
}
```

Users would then be able to catch and inspect the additional information in the error:

```typescript
try {
    const result = await runPulumiCmd(...);
} catch (err) {
    if (err instanceof CommandError) {
        // Use err.result...
        console.log(err.result.stderr);
    }
}
```

This is similar to how `axios` (an http client library) does it: https://github.com/axios/axios#handling-errors

Some more info: https://javascript.info/custom-errors

Reply:
Ah this looks promising. I think we would want to make the command result private in this instance. The Go error design does this so that we can swap out error codes, etc over time as we improve the structured output: 

https://github.com/pulumi/pulumi/blob/master/sdk/go/x/auto/errors.go

For error cases that we want consumers to be able to detect, we'd add public methods on `CommandError` for things like 409 conflict, etc.

Reply:
This has been implemented.